File: src/main/java/ro/ubb/dp1819/lab1/exercises/service/IngredientParserService.java
package ro.ubb.dp1819.lab1.exercises.service;

import ro.ubb.dp1819.lab1.exercises.entity.Ingredient;
import ro.ubb.dp1819.lab1.exercises.exception.InvalidCoffeeException;

import java.util.List;
import java.util.stream.Collectors;

public class IngredientParserService {
    private static Ingredient parseIngredient(String input) {
        String[] tokens = input.split(" ");

        if (tokens.length == 4) {
            return new Ingredient.Builder()
                    .setQuantity(tokens[0])
                    .setUnit(tokens[1])
                    .setIngredient(tokens[2])
                    .setAdjective(tokens[3])
                    .build();
        }
        if (tokens.length == 3) {
            return new Ingredient.Builder()
                    .setQuantity(tokens[0])
                    .setUnit(tokens[1])
                    .setIngredient(tokens[2])
                    .build();
        }
        throw new InvalidCoffeeException("Incorrect number of tokens in input string");
    }

    public static List<Ingredient> parseIngredients(List<String> input) {
        return input.stream().map(IngredientParserService::parseIngredient).collect(Collectors.toList());
    }
}

File: src/main/kotlin/ro/ubb/dp1819/gavrila/andrei/lab1/service/IngredientParserService.kt
package ro.ubb.dp1819.gavrila.andrei.lab1.service

import ro.ubb.dp1819.gavrila.andrei.lab1.coffee.Ingredient

class IngredientParserService {

    private fun readIngredient(line: String): Ingredient {
        val parts = line.split(" ")
        return Ingredient(parts[0].toDouble(), parts[1], parts[2], if (parts.size == 4) parts[3] else null)
    }

    fun readIngredients(lines: List<String>) = lines.map(this::readIngredient)

}